src/worker/typosquat/types.rs

```diff
@@ -0,0 +1,238 @@
+                .collect(),
+        })
+    }
+
+    /// Iterates over the names of the top crates.
+    pub(super) fn iter_names(&self) -> impl Iterator<Item = &str> {
+        self.crates.keys().map(String::as_str)
+    }
+}
+
+impl Corpus for TopCrates {
+    fn contains_name(&self, name: &str) -> typomania::Result<bool> {
+        Ok(self.crates.contains_key(name))
+    }
+
+    fn get(&self, name: &str) -> typomania::Result<Option<&dyn Package>> {
+        Ok(self.crates.get(name).map(|krate| krate as &dyn Package))
+    }
+
+    // We're using the default implementation of `possible_squat`, which will compare the owners of
+    // crates — a new crate won't be considered to be typosquatting a top crate if they share one
+    // or more owners. This prevents noise for things like an author releasing `foo2` for an
+    // existing `foo` crate.
+    //
+    // The one thing typogard-crates did that isn't implemented here is some basic NLP on the crate
+    // descriptions: this relies on the Python spaCy module, which we'd either have to reimplement
+    // the fuzzy matching logic from, or install a Python environment in the background-worker
+    // buildpack. Neither of those are particularly attractive options, so we're just going to
+    // accept a potentially higher level of false positives than we'd otherwise get.
+}
+
+/// A crate and its owners.
+pub(super) struct OwnedCrate {
+    krate: Crate,
+    owners: HashSet<UniqueOwner>,
+}
+
+impl OwnedCrate {
+    /// Hydrates a crate and its owners from the database given the crate name.
+    pub(super) fn from_name(conn: &mut PgConnection, name: &str) -> Result<Self, PerformError> {
+        let krate: Crate = Crate::by_exact_name(name).first(conn)?;
+        let owners = krate
+            .owners(conn)?
+            .into_iter()
+            .map(UniqueOwner::from)
+            .collect();
+
+        Ok(Self { krate, owners })
+    }
+}
+
+impl Package for OwnedCrate {
+    fn authors(&self) -> &dyn AuthorSet {
+        self
+    }
+
+    fn description(&self) -> Option<&str> {
+        self.krate.description.as_deref()
+    }
+
+    fn shared_authors(&self, other: &dyn AuthorSet) -> bool {
+        self.owners.iter().any(|owner| other.contains(&owner.0))
+    }
+}
+
+impl AuthorSet for OwnedCrate {
+    fn contains(&self, author: &str) -> bool {
+        self.owners.contains(author)
+    }
+}
+
+/// A representation of an individual owner that can be compared to other owners to determine if
+/// they represent the same unique user or team that may own one or more crates.
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+struct UniqueOwner(String);
+
+impl UniqueOwner {
+    fn new(id: i32, kind: i32) -> Self {
+        Self(format!("{kind}::{id}"))
+    }
+}
+
+impl From<Owner> for UniqueOwner {
+    fn from(value: Owner) -> Self {
+        Self::new(value.id(), value.kind())
+    }
+}
+
+// Implemented so we can call `contains` above with a `&str`.
+impl Borrow<str> for UniqueOwner {
```

Comment:
out of curiosity, why `Borrow` and not `Deref`?

Reply:
Honestly, because that's how I know how to handle this case (string newtype that you want to be able to use as a hashmap key, but also use `&str` to access).

What does the `Deref` version of this look like?

Reply:
> What does the `Deref` version of this look like?

thanks to https://crates.io/crates/derive-deref it would only be `#[derive(Deref)]`, which would derive an impl that would deref to `&String`.

Reply:
I see the issue now — we need to be able to call `contains` with a `&str`, because that's all the `AuthorSet` and `Package` traits require, but `Deref` would require a `&String`.

I think the `Borrow` is correct here.

Reply:
yeah, just played around with it a bit. looks like you're right 👍